### ISSUE TITLE
[suggestion] Auto-sync dependency when "location" in repositories.manifest is modified

### ISSUE THREAD

**Issue by helmesjo** [build2/build2#178]
## TLDR
_See initial discussion [here](https://github.com/build2-packaging/google-benchmark/pull/5#issuecomment-1008900737)._
If I have a dependency on a library called `convertible`, which is located in repository `https://github.com/helmesjo/convertible.git`, changing which branch it uses currently fetches the package when running `bdep update`, but without an explicit `bdep sync -fur` it continues using the version from the previous branch. 
> `bdep sync -fur` will re-fetch all the repository information (`-f`), and upgrade all the dependencies (`-u`), recursively (`-r`)

Should it perhaps rather (1) sync automatically, or (2) suggest to sync with `[y,n]`?

## Example
Let's say my project is initialized in a configuration. Then I add my new dependency:

```
# manifest
...
depends: convertible ^0.1.0-
```

```
# repositories.manifest
...
:
role: prerequisite
location: https://github.com/helmesjo/convertible.git#main
```

Now I run `bdep update`, it fetches dependencies & builds using `main`. All is fine.
Then, there is a change I want to try out in a new branch, say `important-fix`. So I change to:
```
# repositories.manifest
...
:
role: prerequisite
location: https://github.com/helmesjo/convertible.git#important-fix
```

Running `bdep update` (without first explicitly syncing with `bdep sync -fur`) indicates from the logs that it re-fetches the dependency, but when building it's still using the version from `main`:

`bdep update` after changing branch:
```
...
fetching git:https://github.com/helmesjo/convertible.git#important-fix (prerequisite of dir:path/to/myproject)
querying https://github.com/helmesjo/convertible.git
fetching from https://github.com/helmesjo/convertible.git
...
```

But if my code require something in branch `important-fix` to build, you'll notice that it still fails. I have to explicitly invoke `bdep sync -fur` which then suggests me to upgrade `convertible`. Building after that then works & whatever's in `important-fix` is available.

One easy way to test this is just having another local repository as the `location`. Do init & update, then in said repository make a new branch and commit a change in eg. root `buildfile` adding some garbage that should generate an error. Point to this new branch and run `bdep update` (it won't use the new buildfile with garbage).

**Comment by helmesjo** [comment]
Possible related to this is something I noticed while writing this:

1. If you specify a branch that contains a bad buildfile (eg. just having it contain `asdasdsad` on some line as the latest commit), you'll get this (as expected):
`[...] buildfile:7:10: error: unexpected <newline> after asdasdsad`

2. Then you remove that garbage and make a new commit in the dependee repo.

3. Then, going back to the depender & do `bdep sync -fur`, it will ask if you want to upgrade that dependency, but upon choosing `Y` it instantly fails with the same error:
`[...] buildfile:7:10: error: unexpected <newline> after asdasdsad`

Not even `bdep deinit` & `bdep init` in the depender appears to get around this. I had to do `bdep config remove @myconfig` and re-init with `--wipe` (or manually remove the folder) before it worked.

**Comment by boris-kolpackov** [comment]
Thanks for the report!

Regarding the original issue: while it is valid in `bpkg` to have a built package without a valid location (we call them "orphans"), it does feel that for a `bdep`-managed project (e.g., a project under development), having an orphaned dependency probably means that it's time to upgrade it. At the least we should probably issue some suggestive diagnostics.

Regarding the second issues: it does seem strange for a dependency. Will need to investigate.
